src/platform/plugins/shared/chart_expressions/expression_partition_vis/public/utils/layers/get_color.ts

```diff
@@ -257,18 +235,8 @@ export const getColor = (
     return byDataPalette.getColor(seriesLayers[1].name) || defaultColor;
   }
 
-  if (isTreemapOrMosaicChart(chartType)) {
-    if (layerIndex < columnsLength - 1) {
-      return defaultColor;
-    }
-    // for treemap use the top layer for coloring, for mosaic use the second layer
-    if (seriesLayers.length > 1) {
-      if (chartType === ChartTypes.MOSAIC) {
-        seriesLayers.shift();
-      } else {
-        seriesLayers.pop();
-      }
-    }
+  if (chartType === ChartTypes.MOSAIC && layerIndex < columnsLength - 1) {
+    return defaultColor;
   }
```

Comment:
This was not doing anything good for the treemap. Deleting this like you did fixed the treemap so we now disregard the treemap here.

The only thing we need to handle here is the base case when the are 2-layers, where the first layer should be `defaultColor` and the rest are defined by `byDataPalette.getColor`.

Removing this code also fixes #202816 so I'll add that to the PR description.